src/interface/swarm_boundaries.hpp

```diff
@@ -0,0 +1,199 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_BOUNDARIES_HPP_
+#define INTERFACE_SWARM_BOUNDARIES_HPP_
+
+#include "swarm_device_context.hpp"
+
+namespace parthenon {
+
+class ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION virtual void Apply(const int n, double &x, double &y, double &z,
+                                            // const SwarmDeviceContext &context) = 0;
+                                            const SwarmDeviceContext &context) const {
+    PARTHENON_FAIL("Calling base class");
+  }
+};
+
+class ParticleBoundIX1Periodic : public ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION void Apply(const int n, double &x, double &y, double &z,
+                                    const SwarmDeviceContext &swarm_d) const override {
+    if (x < swarm_d.x_min_global_) {
+      x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);
+    }
+  }
+};
+
+class ParticleBoundIX1Outflow : public ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION void Apply(const int n, double &x, double &y, double &z,
+                                    const SwarmDeviceContext &swarm_d) const override {
+    swarm_d.MarkParticleForRemoval(n);
+  }
+};
+
+class ParticleBoundIX1Reflect : public ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION void Apply(const int n, double &x, double &y, double &z,
+                                    const SwarmDeviceContext &swarm_d) const override {
+    if (x < swarm_d.x_min_global_) {
+      x = swarm_d.x_min_global_ + (swarm_d.x_min_global_ - x);
+    }
+  }
+};
```

Comment:
Where does the velocity get reversed upon reflection?

What happens when a particle exits through a domain edge or corner?

Reply:
Oops, good catch! Yes this is an error. Let me see about fixing this and then adding a unit test for the boundaries...

Reply:
OK this is actually more subtle, because particle velocity isn't a default swarm parameter (only `x`, `y`, and `z` are). So parthenon itself doesn't know which variables to change the direction of. I've removed the default reflecting boundary condition and these will now need to be implemented manually. Reintroducing default reflecting boundaries would probably require adding default velocity variables to `Swarm`s. 

Also, I couldn't find a clear path forward for implementing unit tests of particle boundaries. The `Swarm::ApplyBoundaries_` function requires a large amount of parthenon to be built up because it depends on `bvals`, and I'm not sure just testing the boundary classes themselves is particularly useful since they're so simple (I'd be more worried about how e.g. particles the move across corners are moved back onto the mesh by multiple boundary calls, but this is already captured by the `periodic` boundaries in the particles examples, at least if there are enough particles...)